doc/rdup-up.1.in
'\" t
.TH RDUP-UP 1 "13 Dec 2008" "@PACKAGE_VERSION@" "@PACKAGE_NAME@"
.SH NAME
@PACKAGE_NAME@-up \- update a directory tree with a rdup archive
.SH SYNOPSIS
.B @PACKAGE_NAME@-up
[\fIOPTION\fR]...
\fIDIRECTORY\fR

.SH DESCRIPTION
With \fBrdup-up\fR you can update an (possibly) existing directory
structure with a rdup archive.

The rdup archive has to be given to \fBrdup-up\fR's standard input.

.SS Username and uids
\fBrdup\fR outputs both the user name and uid, the receiving system (which
may be a totally different system) checks if the user name and uid
match. If the user name and uid don't match the (numeric) uid is
used on the file. The same holds true for the group name and gid.

.SS File ownership
As \fBrdup\fR supports backups via SSH the following situation can occur:
locally \fBrdup\fR is run a root, but \fBrdup-up\fR is run as a non-root
user (the one logged in via SSH). In this case the original owner- and
group name can not be set. If this happens \fBrdup-up\fR will create
a \fI._rdup_.\fR file which contains the user/group information, also
see the \fI\-u\fR flag for \fBrdup-tr\fR (and \fBrdup\fR).

.SH OPTIONS
.TP
.B \-n
Do a dry-run and do not create anything on disk.
.TP
.B \-t 
Create DIRECTORY (ala mkdir -p) if it does not exist.
.TP
.B \-s N
Strip N path components from a pathname. If the resulting pathname is
empty after this operation it is skipped. Be careful however with the following
structure:

    /foo
    /foo/bar
    /foo/bar/bla.txt
    /foo/blork/bla.txt


With \fBrdup-up -s\fI2\fR this will leave:


    <empty>
    <empty>
    /bla.txt
    /bla.txt


And the last 'bla.txt' will \fIover write\fR the previous one, this will
happen without warnings.
.TP
.B \-r PATH
This option is related to the \-s option, but works different. The
string PATH is removed from (the beginning of) each pathname. With \fI-r
/home/backup\fR the pathname \fI/home/backup/bin/mycmd\fR becomes
\fI/bin/mycmd\fR. The same could be done with \fB-s 2\fR, but then you need
to count the slashes. Note \-s is always performed before \-r.
.TP
.B \-v
Be more verbose and echo the processed files to \fIstandard output\fR.
.TP
.B \-T
Show a table of contents of the rdup stream received (ala tar \-tf \-).
With \-T the directory argument is optional. \-T unsets any verbose (\-v) options.
.TP
.B \-u
Do not create a \fI._rdup_.\fR file which contains user/group information when failing
to \fBchown\fR the actual file or directory. Useful when restoring a backup when you
do not want to see \fI._rdup._\fR-files being created.
.TP
.B \-q
Silence 'chown' failures even when running as root. This can be helpful when the 
file system does not implement 'chown' or disallows it ('sshfs' for instance).
.TP
.B \-h
A short help message.
.TP
.B \-V
Show the version.

.SH EXIT CODE
\fBrdup-up\fR return a zero exit code on success, otherwise 1 is returned.

.SH AUTHOR
Written by Miek Gieben. 

.SH REPORTING BUGS
Report bugs to <miek@miek.nl>. 

.SH SEE ALSO
http:/www.miek.nl/projects/rdup/ is the main site of rdup. Also see rdup(1),
rdup-tr(1) and rdup-backups(7).

.SH COPYRIGHT
Copyright (C) 2005-2010 Miek Gieben. This is free software. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.
.PP
Licensed under the GPL version 3. See the file LICENSE in the source distribution
of rdup.


ChangeLog
26 May 2012: 1.1.14 Miek Gieben <miek@miek.nl>
   * [rdup] check for the existence of ._rdup_. files and parse its contents
     to restore the actual files' ownership (Miek Gieben)
   * [rdup-up] save user/group info in ._rdup_. files when chown(1)ing has failed.
     Useful when backing up via sshfs. See rdup(1) "Non-root backups" (Miek Gieben)
   * [rdup-up] add -u flag to suppress creating these ._rdup_. files. 
     (useful when restoring) (Miek Gieben)
   * [rdup] add -u flag to suppress interpreting these ._rdup_. files. (Miek Gieben)
   * [rdup-tr] compile fix when compiling with gcc 4.7 (reported by D. L.)


18 Feb 2012: 1.1.13 Miek Gieben <rdup@miek.nl>
   * Make rdup-tr compile when libnettle was not found (Miek Gieben)
   * Fix for very large inodes and filesizes (Kojedzinszky Richard)
   * Fix tar/pax output when the filesize of files changes (rdup -Pgzip for
     instance) (Reported by: Alejandro Pulver). This created files padded
     with zero bytes.


21 Jan 2012: 1.1.12 Miek Gieben <rdup@miek.nl>
   * Test suite fixed
   * Fix Debian bug 627617


18 Feb 2011. 1.1.11 Miek Gieben <rdup@miek.nl>
   * Polish
   * rdup-up: added -q option. Be quiet when chown
     fails even when root (handy for backups on sshfs)


05 Dec 2010: 1.1.10 Miek Gieben <rdup@miek.nl>
   * One memleak fix removed (Tom Hendrikx)


04 Dec 2010: 1.1.9 Miek Gieben <rdup@miek.nl>
   * Fix whooping amount of memleaks (rdup-tr, rdup-up) (Crispin Boylan)


24 Oct 2010: 1.1.8 Miek Gieben <rdup@miek.nl>
   * rdup: Check for two more I/O errors when reading files
   * rdup: Permission fix for AIX (Crispin Boylan)
   * autoconf: Fix dirfd detection on mac OS X (still does not work)


21 Jun 2010: 1.1.7 Miek Gieben <rdup@miek.nl>
    * rdup-up: also create named pipes (fifo)


18 May 2010: 1.1.6 Miek Gieben <rdup@miek.nl>
    * rdup-simple: cleanups + portability


09 Apr 2010: 1.1.5 Miek Gieben <rdup@miek.nl>
    * compile fix


04 Apr 2010: 1.1.4 Miek Gieben <rdup@miek.nl>
    * re-touch timestamp after the backup is made when -a is used,
      otherwise it is done before the backup
    * spring cleaning, remove/clean BUGBUGs and some dead code and
      polish the documentation a bit
    * revert changes in mkpath.c to make FreeBSD happy
    * fix hardlinks in tar/pax output (they must come last)


22 Mar 2010: 1.1.3 Miek Gieben <rdup@miek.nl>
    * rdup: fix reset atime (-a)
    * rdup-simple: add -a (reset atime flag)
    * documentation bugfixes


14 Mar 2010: 1.1.2 Miek Gieben <rdup@miek.nl>
    * rdup: remove sig_atomic_t (build fix for ubuntu 8.04)
     (reported by Andrew Ittner)
    * rdup-up: fixed nasty bug in dir.c dir_parent()
     (leave caller's string alone)
    * rdup-up: rm.c: also use dir_parent()
    * tests: a 'fail'ing gpg test is now 'unresolved'
    * Added contrib/ directory and added prune.rb script
      (prune.rb provided by Matthias-Christian Ott)


2 Mar 2010: 1.1.1 Miek Gieben <rdup@miek.nl>
    * tests: Skip gpg test when gpg is not found (Tom)
    * rdup: Rename strmode() to rdup_strmode for FreeBSD (Charlie Kester)
    * rdup: Dont descend when toplevel directory is a symlink (same behavoir
      as Gnu tar) (Reported by Charlie Kester)
    * rdup: reap our childeren with waitpid() (Oliver Dain)
    * build: Building fix on rpm distros (SuSE/Fedora) (Sven Hartrumpf)
    * build: add --enable-debug to build debug build


24 Jan 2010: 1.1.0 Miek Gieben <rdup@miek.nl>
    * Release!
    * Tweaks and merges from the master branch
    * Remember: installs in /usr/local


22 Dec 2009: 1.1.0-rc1 Miek Gieben <rdup@miek.nl>
    * First release of the 1.1.0 branch
    * format change: %U, %G and %t added
    * default install path: /usr/local/
    * rdup: -a restore atime when reading files
    * rdup: -P is moved from rdup-tr 
    Todo:
    * documentation
    * more testing


14 Dec 2009: 1.1.0-dev1 Miek <rdup@miek.nl>
    * rdup added:
	- -a: reset atime
    * rdup-up:
	- set mtime
    * revert back to libnettle
	- and back to AES encryption


09 Aug 2009: 1.1.0-dev0 Miek <rdup@miek.nl>
    * Format change: add username, groupname and mtime
      Always output the files' contents
      %U username
      %G groupname
      %t is mtime of the fs object
    * Adapt all utils to this format change
	- rdup now handles -P itself
	- rdup is the only tool that accesses the files
	- rdup-tr does not do -P anymore
	- rdup-tr only processes from stdin to stdout
	- rdup-up only processes stdin
    * Loads of code cleanups
    * Use OpenSSL in favor of libnettle
	- AES encryption is replaced by blowfish
    * Change default install path /usr/local (Sebastian Kügler)


rdup-up.c
/*
 * Copyright (c) 2009 - 2011 Miek Gieben
 * See LICENSE for the license
 * rdup-up -- update an directory tree with
 * and rdup archive
 */

#include "rdup-up.h"
#include "io.h"

char *PROGNAME="rdup-up";
/* options */
gint opt_verbose 	   = 0;                         /* be more verbose */
gint opt_output		   = O_RDUP;			/* set these 2 so we can use parse_entry */
gint opt_input	           = I_RDUP;
gboolean opt_quiet         = FALSE;                     /* don't complain about chown() errors */
gboolean opt_dry	   = FALSE;			/* don't touch the filesystem */
gboolean opt_table	   = FALSE;			/* table of contents */
gboolean opt_top	   = FALSE;			/* create top dir if it does not exist */
gboolean opt_chown         = TRUE;                      /* create ._rdup._ files with user/group info */
gchar *opt_path_strip	   = NULL;			/* -r PATH, strip PATH from pathnames */
guint opt_path_strip_len   = 0;				/* number of path labels in opt_path_strip */
guint opt_strip		   = 0;				/* -s: strippath */
int sig           = 0;
GSList *hlink_list		   = NULL;			/* save hardlink for post processing */		
extern int opterr;
int opterr		   = 0;

/* update the directory with the archive */
static gboolean
update(char *path)
{
	struct rdup    *rdup_entry;
	size_t         line, i, pathsize;
	size_t	       pathlen;
	char           *buf, *pathbuf, *n, *p;
	char           delim;
	gboolean       ok;
	GHashTable     *uidhash; /* holds username -> uid */
	GHashTable     *gidhash; /* holds groupname -> gid */

	buf	= g_malloc(BUFSIZE + 1);
	pathbuf = g_malloc(BUFSIZE + 1);
	i       = BUFSIZE;
	delim   = '\n';
	line    = 0;
	ok      = TRUE;
	uidhash = g_hash_table_new(g_str_hash, g_str_equal);
	gidhash = g_hash_table_new(g_str_hash, g_str_equal);
	if (path)
		pathlen = strlen(path);
	else
		pathlen = 0;

	while ((rdup_getdelim(&buf, &i, delim, stdin)) != -1) {
		line++;
		n = strrchr(buf, '\n');
		if (n)
			*n = '\0';

		if (!(rdup_entry = parse_entry(buf, line))) {
			/* msgs from entry.c */
			exit(EXIT_FAILURE);
		}

		/* we have a valid entry, read the filename */
		pathsize = fread(pathbuf, sizeof(char), rdup_entry->f_name_size, stdin);

		if (pathsize != rdup_entry->f_name_size) {
			msg(_("Reported name size (%zd) does not match actual name size (%zd)"),
					rdup_entry->f_name_size, pathsize);
			exit(EXIT_FAILURE);
		}
		pathbuf[pathsize] = '\0';
		if (pathbuf[0] != '/') {
			msg(_("Pathname does not start with /: `%s\'"), pathbuf);
			exit(EXIT_FAILURE);
		}

		rdup_entry->f_name = pathbuf;

		/* extract target from rdup_entry */
		if (S_ISLNK(rdup_entry->f_mode) || rdup_entry->f_lnk) {
			// filesize is spot where to cut
			rdup_entry->f_name[rdup_entry->f_size] = '\0';
			rdup_entry->f_target = rdup_entry->f_name +
				rdup_entry->f_size + 4;
		} else {
			rdup_entry->f_target = NULL;
		}

		/* strippath must be inserted here */
		if (opt_strip)
			strippath(rdup_entry);

		if (opt_path_strip)
			strippathname(rdup_entry);

		if (!rdup_entry->f_name)
			p = NULL;
		else {
			/* avoid // at the beginning */
			if ( (pathlen == 1 && path[0] == '/') || pathlen == 0 ) {
				p = g_strdup(rdup_entry->f_name);
			} else {
				p = g_strdup_printf("%s%s", path, rdup_entry->f_name);
				rdup_entry->f_name_size += pathlen;
				if (S_ISLNK(rdup_entry->f_mode) || rdup_entry->f_lnk)
					rdup_entry->f_size += pathlen;
			}
		}
		rdup_entry->f_name = p;
		if (mk_obj(stdin, path, rdup_entry, uidhash, gidhash) == FALSE)
			ok = FALSE;
                g_free(rdup_entry->f_user);
                g_free(rdup_entry->f_group);
		g_free(rdup_entry);
	}

	/* post-process hardlinks */
	if (mk_hlink(hlink_list) == FALSE)
		ok = FALSE;

	g_free(buf);
	g_free(pathbuf);
	return ok;
}

int
main(int argc, char **argv)
{
	struct sigaction sa;
	char		 pwd[BUFSIZE + 1];
	int		 c;
	char		 *path;
	
#ifdef ENABLE_NLS
	if (!setlocale(LC_MESSAGES, ""))
		msg(_("Locale could not be set"));

	bindtextdomain(PACKAGE_NAME, LOCALEROOTDIR);
	(void)textdomain(PACKAGE_NAME);
#endif /* ENABLE_NLS */
	
	/* setup our signal handling */
	sa.sa_flags   = 0;
	sigfillset(&sa.sa_mask);

	sa.sa_handler = got_sig;
	sigaction(SIGPIPE, &sa, NULL);
	sigaction(SIGINT, &sa, NULL);

	if (((getuid() != geteuid()) || (getgid() != getegid()))) {
		msg(_("Will not run suid/sgid for safety reasons"));
		exit(EXIT_FAILURE);
        }

	if (!getcwd(pwd, BUFSIZE)) {
		msg(_("Could not get current working directory"));
		exit(EXIT_FAILURE);
	}

	for(c = 0; c < argc; c++) {
		if (strlen(argv[c]) > BUFSIZE) {
			msg(_("Argument length overrun"));
			exit(EXIT_FAILURE);
		}
	}

	while ((c = getopt (argc, argv, "thnVvus:r:Tq")) != -1) {
		switch (c) {
			case 'v':
				opt_verbose = 1;
				break;
			case 'h':
				usage_up(stdout);
				exit(EXIT_SUCCESS);
			case 'n':
				opt_dry = TRUE;
				break;
			case 'T':
				opt_table = TRUE;
				opt_dry = TRUE;
				opt_verbose = 0;
				break;
                        case 'u':
                                opt_chown = FALSE;
                                break;
			case 's':
                                opt_strip = abs(atoi(optarg));
                                break;
			case 'r':
				/* expand relative paths */
				if (!g_path_is_absolute(optarg))
					opt_path_strip = abspath(g_strdup_printf("%s/%s", pwd, optarg));
				else
					opt_path_strip = abspath(optarg);

				if (!opt_path_strip) {
					msg(_("Failed to expand path `%s\'"), optarg);
					exit(EXIT_FAILURE);
				}

				if (opt_path_strip[strlen(opt_path_strip) - 1] != '/')
					opt_path_strip = g_strdup_printf("%s/", opt_path_strip);

				/* count the number of labels */
				guint i;
				for(i = 0; i < strlen(opt_path_strip); i++) {
					if (opt_path_strip[i] == '/')
						opt_path_strip_len++;
				}
				opt_path_strip_len--;	/* we added the closing slash, so need -1 here */
				break;
                        case 'q':
                                opt_quiet = TRUE;
                                break;
			case 't':
				opt_top = TRUE;
				break;
			case 'V':
#ifdef DEBUG
                                fprintf(stdout, "%s %s (with --enable-debug)\n", PROGNAME, VERSION);
#else
                                fprintf(stdout, "%s %s\n", PROGNAME, VERSION);
#endif /* DEBUG */
				exit(EXIT_SUCCESS);
			default:
				msg(_("Unknown option seen `%c\'"), optopt);
				exit(EXIT_FAILURE);
		}
	}
	argc -= optind;
	argv += optind;

	if (opt_table) {
		path = NULL;
	} else {
		if (argc != 1) {
			msg(_("Destination directory is required"));
			exit(EXIT_FAILURE);
		}
		if (!g_path_is_absolute(argv[0])) {
			gchar* full_path = g_strdup_printf("%s/%s", pwd, argv[0]);
			path = abspath(full_path);
			g_free(full_path);
		}
		else
			path = abspath(argv[0]);
	}

	if (!opt_dry) {
		if (opt_top) {
			/* this is not 100%, but better than nothing */
			if (g_file_test(path, G_FILE_TEST_IS_REGULAR)) {
				msg(_("Failed to create directory `%s\'"), path);
				exit(EXIT_FAILURE);
			}
			if (mkpath(path, 00777) == -1) {
				msg(_("Failed to create directory `%s\': %s"), path, strerror(errno));
				exit(EXIT_FAILURE);
			}
		} else {
			if (!g_file_test(path, G_FILE_TEST_IS_DIR)) {
				msg(_("No such directory: `%s\'"), path);
				exit(EXIT_FAILURE);
			}
		}
	}

	if (update(path) == FALSE)
		exit(EXIT_FAILURE);

	g_free(path);

	exit(EXIT_SUCCESS);
}
